OSL Misc work

est77 opened the pull request:
Added OSL's coordinate systems transforms from object, camera and world to NDC and raster spaces.

OSL's SurfaceShaders:
A surface shader that runs an OSL shader to post-process the shading results. Might be useful to do color corrections and other special effects.
The commit includes useful things, like marking OSL shader parameters as outputs and getting shader symbols to be able to fetch results from them.

dictoon commented on src/appleseed/renderer/kernel/rendering/rendererservices.cpp:
```diff
@@ -347,6 +354,72 @@ bool RendererServices::transform_points(
+        if (from == g_world_ustr || from == g_common_ustr || from == g_shader_ustr)
+        {
+            // Copy the points to Pout.
+            for (size_t i = 0; i < npoints; ++i)
```
Signed/unsigned mismatch here (i is `size_t`, npoints is `int`).

dictoon commented on src/appleseed/renderer/kernel/shading/shadingpoint.h:
```diff
@@ -259,6 +259,7 @@ class ShadingPoint
     friend class CurveLeafVisitor;
     friend class Intersector;
     friend class OSLShaderGroupExec;
+    friend class RendererServices;
```
Order alphabetically.

dictoon commented on src/appleseed/renderer/modeling/shadergroup/shadergroup.h:
```diff
@@ -134,6 +134,10 @@ class APPLESEED_DLLSYMBOL ShaderGroup
     // Return a reference-counted (but opaque) reference to the internal OSL shader group.
     OSL::ShaderGroupRef& shader_group_ref() const;
 
+    // OSL Symbols
```
`// OSL symbols.`

dictoon commented on src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp:
```diff
@@ -0,0 +1,246 @@
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2015 Esteban Tovagliari, The appleseedhq Organization
```
Fix year in copyright notice (it probably should say 2016, or 2015-2016, until I update all notices for the year 2017).

dictoon commented on src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp:
```diff
@@ -0,0 +1,246 @@
+        }
+
+        virtual void evaluate(
+            SamplingContext&        sampling_context,
```
Align all arguments vertically.

dictoon commented on src/appleseed/renderer/modeling/surfaceshader/oslsurfaceshader.cpp:
```diff
@@ -0,0 +1,246 @@
+        {
+            if (m_shader_group)
+            {
+                const OSL::ShaderSymbol *color_sym =
```
Star goes on the type's side.

dictoon commented:
Pretty cool!

Can you provide an example of a postprocessing OSL surface shader?

est77 commented:
Fixed and rebased.
I don't have any example now of OSL surface shaders. I'll see if I can borrow some asset and make one.
I was more interested in merging the code that deals with OSL symbols and shader outputs.
We'll need it for light attenuation shaders, AOVs and possibly other uses.

dictoon commented:
I suggest adding a few test scenes using very simple OSL surface shaders (e.g. a surface shader that does nothing, or just multiplies RGB by 2, or something like that) to make sure this code works and continues to do so in the future.

est77 commented:
Ok. I'll add some test scenes in a following PR.
